tools/kvmexit.py

```diff
@@ -0,0 +1,293 @@
+#!/usr/bin/env python
+#
+# kvmexit.py
+#
+# Display the exit_reason and its statistics of each kvm_exit
+# for all vcpus of all virtual machines. For example,
+# $./kvmexit.py
+#  TGID     PID      KVM_EXIT_REASON                     STAT
+#  1273551  1273568  EXIT_REASON_MSR_WRITE               6
+#  1274253  1274261  EXIT_REASON_EXTERNAL_INTERRUPT      1
+#  1274253  1274261  EXIT_REASON_HLT                     12
+#  ...
+#
+# @TGID: each vitual machine's pid in the user space.
+# @PID: the user space's thread of each vcpu of that virtual machine.
+# @KVM_EXIT_REASON: the reason why the vm exits.
+# @STAT: the counts of the @KVM_EXIT_REASON since the vm starts, and
+#        it is recorded every second by setting `sleep(1)`.
+#
+# REQUIRES: Linux 4.7+ (BPF_PROG_TYPE_TRACEPOINT support)
+#
+# Copyright (c) 2019 BYTEDANCE Inc.
+#
+# Author(s):
+#   Fei Li <lifei.shirley@bytedance.com>
+
+
+from __future__ import print_function
+from time import sleep, strftime
+from bcc import BPF
+import argparse
+import multiprocessing
+import signal
+import subprocess
+
+#
+# Process Arguments
+#
+
+# arg validation
+def positive_int(val):
+    try:
+        ival = int(val)
+    except ValueError:
+        raise argparse.ArgumentTypeError("must be an integer")
+
+    if ival < 0:
+        raise argparse.ArgumentTypeError("must be positive")
+    return ival
+
+def positive_nonzero_int(val):
+    ival = positive_int(val)
+    if ival == 0:
+        raise argparse.ArgumentTypeError("must be nonzero")
+    return ival
+
+# arguments
+examples = """examples:
+    ./kvmexit             # Display kvm_exit_reason and its statistics until Ctrl-C
+    ./kvmexit 5           # Display kvm_exit_reason and its statistics for 5 seconds only
+"""
+parser = argparse.ArgumentParser(
+    description="Display kvm_exit_reason and its statistics at a timed interval",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+parser.add_argument("duration", nargs="?", default=99999999,
+    type=positive_nonzero_int,
+    help="duration of trace, in seconds")
+
+args = parser.parse_args()
+duration = int(args.duration)
+
+#
+# Setup BPF
+#
+
+# load BPF program
+bpf_text = """
+#include <linux/delay.h>
+
+#define REASON_NUM 69
+#define TGID_NUM 1024
+
+struct exit_count {
+    u64 exit_ct[REASON_NUM];
+};
+BPF_PERCPU_ARRAY(init_value, struct exit_count, 1);
+BPF_TABLE("percpu_hash", u64, struct exit_count, pcpu_kvm_stat, TGID_NUM);
+
+struct cache_info {
+    u64 cache_pid_tgid;
+    struct exit_count cache_exit_ct;
+};
+BPF_PERCPU_ARRAY(pcpu_cache, struct cache_info, 1);
+
+FUNC_ENTRY {
+    u64 cur_pid_tgid = bpf_get_current_pid_tgid();
+    int zero = 0;
+    int cache_miss = 0;
+    u32 er = GET_ER;
+    if (er >= REASON_NUM) {
+        return -1;
```

Comment:
Just return 0 here. For raw_tracepoint, return value is not used. for tracepoint, return 1 is different from return 0. return 1 permits kernel perf event output continues, this is important when you enable debugfs tracepoint. but I prefers to return 0 in this function for such complexity. It is unlikely people will use debugfs tracepoint if this tool is used.

Reply:
> Just return 0 here. For raw_tracepoint, return value is not used. for tracepoint, return 1 is different from return 0. return 1 permits kernel perf event output continues, this is important when you enable debugfs tracepoint. but I prefers to return 0 in this function for such complexity. It is unlikely people will use debugfs tracepoint if this tool is used.

Would you mind pointing the code saying "return 1 permits kernel perf event output"?  BTW, will "return -1" bother this logic?

Reply:
In kernel/events/core.c, we have
```
  struct trace_event_call *call, u64 count,
                               struct pt_regs *regs, struct hlist_head *head,
                               struct task_struct *task)
{
        if (bpf_prog_array_valid(call)) {
                *(struct pt_regs **)raw_data = regs;
                if (!trace_call_bpf(call, raw_data) || hlist_empty(head)) {
                        perf_swevent_put_recursion_context(rctx);
                        return;
                }
        }
        perf_tp_event(call->event.type, count, raw_data, size, regs, head,
                      rctx, task);
}
EXPORT_SYMBOL_GPL(perf_trace_run_bpf_submit);
```

If bpf program returns none zero, it is possible control may go to perf_tp_event(). But this is for regular tracepoint, previously I probably didn't pay attention to the fact that it used raw tracepoint here.

For raw tracepoint, the return value is ignored. See c4f6699dfcb85 ("bpf: introduce BPF_RAW_TRACEPOINT"). So let us just return 0 in all places.